database/manager/manager.go

```
@@ -0,0 +1,238 @@
+func New(dbDirPath string, currentVersion version.Version) (Manager, error) {
+	parser := version.NewDefaultParser()
+
+	currentDBPath := path.Join(dbDirPath, currentVersion.String())
+	currentDB, err := leveldb.New(currentDBPath, 0, 0, 0)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't create db at %s: %w", currentDBPath, err)
+	}
+
+	semDBs := make([]*SemanticDatabase, 1)
+	semDBs[0] = &SemanticDatabase{
+		Database: currentDB,
+		Version:  currentVersion,
+	}
+	err = filepath.Walk(dbDirPath, func(path string, info os.FileInfo, err error) error {
+		// Skip the root directory
+		if path == dbDirPath {
+			return nil
+		}
+
+		// The database directory should only contain database directories, no files.
+		if !info.IsDir() {
+			return fmt.Errorf("unexpectedly found non-directory at %s", path)
+		}
+		_, dbName := filepath.Split(path)
+		version, err := parser.Parse(dbName)
+		if err != nil {
+			return err
+		}
+
+		// If [version] is greater than or equal to the specified version
+		// skip over creating the new database to avoid creating the same db
+		// twice or creating a database with a version ahead of the desired one.
+		if cmp := version.Compare(currentVersion); cmp >= 0 {
+			return filepath.SkipDir
+		}
+
+		db, err := leveldb.New(path, 0, 0, 0)
+		if err != nil {
+			return fmt.Errorf("couldn't create db at %s: %w", path, err)
+		}
+
+		semDBs = append(semDBs, &SemanticDatabase{
+			Database: db,
+			Version:  version,
+		})
+
+		return filepath.SkipDir
+	})
+
+	SortDescending(semDBs)
+
+	m := &manager{
+		databases: semDBs,
+	}
+
+	// If an error occurred, close all of the opened databases
+	// and return the original error
+	if err != nil {
+		_ = m.Close()
+		return nil, err
+	}
+
+	return m, nil
+}
+
+// NewPrefixDBManager creates a new manager with each database instance prefixed
+// by [prefix]
+func (m *manager) NewPrefixDBManager(prefix []byte) Manager {
+	m, _ = m.wrapManager(func(sdb *SemanticDatabase) (*SemanticDatabase, error) {
```

Should we discard this error?

---

Yes. wrapManager will only return an error if the passed in function does, since the function passed in never returns an error, this error can be safely ignored.